### ISSUE TITLE
请问generate_cover_frames_lmdb.py会直接从视频里抽帧来构造lmdb，那为什么我们还要下载帧数的压缩包呢，能否直接提供视频的压缩包，谢谢

### ISSUE THREAD

**Issue by NKUShaw** [westlake-repl/MicroLens#27]

**Comment by yxin98** [comment]
@NKUShaw 你好，视频和帧数的压缩包在resources都提供了，比如“https://recsys.westlake.edu.cn/MicroLens-100k-Dataset/”里面。至于现成的LMDB，这个无法提供，它的构造会使得它比输入数据（比如所有视频帧）还大

**Comment by NKUShaw** [comment]
> [@NKUShaw](https://github.com/NKUShaw) 你好，视频和帧数的压缩包在resources都提供了，比如“[https://recsys.westlake.edu.cn/MicroLens-100k-Dataset/”里面。至于现成的LMDB，这个无法提供，它的构造会使得它比输入数据（比如所有视频帧）还大](https://recsys.westlake.edu.cn/MicroLens-100k-Dataset/%E2%80%9D%E9%87%8C%E9%9D%A2%E3%80%82%E8%87%B3%E4%BA%8E%E7%8E%B0%E6%88%90%E7%9A%84LMDB%EF%BC%8C%E8%BF%99%E4%B8%AA%E6%97%A0%E6%B3%95%E6%8F%90%E4%BE%9B%EF%BC%8C%E5%AE%83%E7%9A%84%E6%9E%84%E9%80%A0%E4%BC%9A%E4%BD%BF%E5%BE%97%E5%AE%83%E6%AF%94%E8%BE%93%E5%85%A5%E6%95%B0%E6%8D%AE%EF%BC%88%E6%AF%94%E5%A6%82%E6%89%80%E6%9C%89%E8%A7%86%E9%A2%91%E5%B8%A7%EF%BC%89%E8%BF%98%E5%A4%A7)

我没有看到视频的压缩包，只看到视频的文件夹。我遇到的问题是要走你们的处理流程是输入视频给[generate_cover_frames_lmdb.py](https://github.com/westlake-repl/MicroLens/blob/master/Data%20Processing/generate_cover_frames_lmdb.py)然后得到frames，再对frames处理得到lmdb。。。问题是我现在没有视频。。。这就尴尬了，我自己修改了代码直接读取你们提供的frames，结果现在跑不通baseline实验各种报错，我不清楚到底是我处理的lmdb文件有问题还是什么有问题。

**Comment by yxin98** [comment]
@NKUShaw 你好，我们提供了frames的压缩包，直接用frames加上lmdb的处理代码，可以得到LMDB文件。先确保本地正确生成LMDB文件（这个文件很大，运行要一段时间），然后测试下lmdb读取是否成功。

**Comment by yxin98** [comment]
@NKUShaw 你好，我看到你说的了，你应该是看到
“pretrain_path = './videomae_base' # to input
video_path = './xxx_videos'”被误导了。

第一个你需要填你的VideoMAEFeatureExtractor下载后的保存路径，第二个你改成xxx_frames就好，它的原理是基于frames，实际不是基于每个视频。我们说的video，默认是frames（视频学习都是基于多个帧的，一般没有把视频作为输入的，输入视频中间也是抽帧来学习，没有直接对视频进行学习的）

**Comment by NKUShaw** [comment]
```
def generate_video_lmdb(pretrain_path, video_path, frame_no):

	configuration = VideoMAEConfig()
	feature_extractor = VideoMAEFeatureExtractor(configuration)
	feature_extractor = feature_extractor.from_pretrained(pretrain_path)

	frames_folder = 'video_frames/' # a folder that restores frames of videos
	# {video1-1.jpg, video1-2.jpg, video1-3.jpg, video1-4.jpg, video1-5.jpg} refer to 5 frames extracted from video1
	lmdb_path = 'video_frames.lmdb' # to input

	all_video = [x.replace('.mp4', '.jpg') for x in os.listdir(video_path)]
	video_num = len(all_video)
	isdir = os.path.isdir(lmdb_path)
	lmdb_env = lmdb.open(lmdb_path, subdir=isdir, map_size=video_num * np.zeros((3, 224, 224)).nbytes*10,
		readonly=False, meminit=False, map_async=True)
	txn = lmdb_env.begin(write=True)
	write_frequency = 100
	for idx, video in enumerate(tqdm.tqdm(all_video)):
		temp = []
		video_id = int(video)
		for frame_idx in range(1, frame_no+1):
			img = np.array(transform(Image.open(frames_folder+str(video_id)+'-'+str(frame_idx)+'.jpg').convert('RGB')))		
			temp.append(img)
		temp = list(np.array(temp))
		temp = feature_extractor(temp, return_tensors="pt").pixel_values[0]
		temp = LMDB_VIDEO(np.array(temp))
		txn.put(u'{}'.format(video_id).encode('ascii'), pickle.dumps(temp))
		if video_id % write_frequency == 0 and idx != 0:
			txn.commit()
			txn = lmdb_env.begin(write=True)
	txn.commit()
	keys = [u'{}'.format(k).encode('ascii') for k in range(video_num)]
	with lmdb_env.begin(write=True) as txn:
		txn.put(b'__keys__', pickle.dumps(keys))
		txn.put(b'__len__', pickle.dumps(len(keys)))
	print(len(keys))
	print("Flushing database ...")
	lmdb_env.sync()
	lmdb_env.close()

generate_image_lmdb()

pretrain_path = 'MCG-NJU/videomae-base'
video_path = '../Data/MicroLens-50k-Dataset/MicroLens-50k_frames_interval_1_number_5'
for frame_no in [1, 2, 3, 4, 5]:
	generate_video_lmdb(pretrain_path, video_path, frame_no)

```

我疑惑的点就是frames_folder = 'video_frames/' 和video_path这里有什么区别，
还有lmdb_path = 'video_frames.lmdb'这里，看起来for循环for frame_no in [1, 2, 3, 4, 5]是会重复填充这个lmdb文件，还是说需要分别建立五个lmdb文件，但是在你们的baseline代码只需要读取一个lmdb文件

**Comment by yxin98** [comment]
@NKUShaw ，刚才说错了，frames_folder才是存放frames的地方，至于video_path，实际上可以去掉这个参数。video_path 实际从未被真正用来读取视频文件内容。这是新的函数，它不需要定义video_path也能运行。

def generate_video_lmdb(pretrain_path, frame_no):
    from transformers import VideoMAEFeatureExtractor, VideoMAEConfig
    import os, lmdb, pickle, numpy as np, tqdm
    from PIL import Image
    import torchvision.transforms as transforms

    # 初始化 VideoMAE 特征提取器
    configuration = VideoMAEConfig()
    feature_extractor = VideoMAEFeatureExtractor(configuration)
    feature_extractor = feature_extractor.from_pretrained(pretrain_path)

    # 帧图文件夹
    frames_folder = 'video_frames/'  # 存放帧图像的目录
    lmdb_path = 'video_frames.lmdb'  # 输出 LMDB 文件名

    # 从帧文件名中推断出视频 ID
    all_frames = os.listdir(frames_folder)
    # 例如 video_frames/1-1.jpg → 提取前面的 1
    video_ids = sorted(list({f.split('-')[0] for f in all_frames if f.endswith('.jpg')}))
    video_num = len(video_ids)

    # 图像标准化
    transform = transforms.Compose([
        transforms.Resize((224, 224)),
        transforms.ToTensor(),
        transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5)),
    ])

    # 定义 LMDB 数据类
    class LMDB_VIDEO:
        def __init__(self, video):
            self.video = video.tobytes()

    # 创建 LMDB 数据库
    isdir = os.path.isdir(lmdb_path)
    lmdb_env = lmdb.open(
        lmdb_path, subdir=isdir,
        map_size=video_num * np.zeros((3, 224, 224)).nbytes * 10,
        readonly=False, meminit=False, map_async=True
    )
    txn = lmdb_env.begin(write=True)
    write_frequency = 100

    for idx, vid in enumerate(tqdm.tqdm(video_ids)):
        temp = []
        video_id = int(vid)
        for frame_idx in range(1, frame_no + 1):
            frame_path = os.path.join(frames_folder, f"{video_id}-{frame_idx}.jpg")
            if not os.path.exists(frame_path):
                continue
            img = np.array(transform(Image.open(frame_path).convert('RGB')))
            temp.append(img)

        if not temp:
            continue

        temp = list(np.array(temp))
        temp = feature_extractor(temp, return_tensors="pt").pixel_values[0]
        temp = LMDB_VIDEO(np.array(temp))
        txn.put(u'{}'.format(video_id).encode('ascii'), pickle.dumps(temp))

        if video_id % write_frequency == 0 and idx != 0:
            txn.commit()
            txn = lmdb_env.begin(write=True)

    # 提交并保存键索引
    txn.commit()
    keys = [u'{}'.format(k).encode('ascii') for k in range(video_num)]
    with lmdb_env.begin(write=True) as txn:
        txn.put(b'__keys__', pickle.dumps(keys))
        txn.put(b'__len__', pickle.dumps(len(keys)))

    print(len(keys))
    print("Flushing database ...")
    lmdb_env.sync()
    lmdb_env.close()
